File: src/main/java/land/leets/global/filter/ExceptionHandleFilter.java
package land.leets.global.filter;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import land.leets.global.error.ErrorCode;
import land.leets.global.error.ErrorResponse;
import land.leets.global.error.exception.ServiceException;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class ExceptionHandleFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        try {
            filterChain.doFilter(request, response);
        } catch (ServiceException e) {
            sendErrorResponse(response, e.getErrorCode());
        } catch (Exception e) {
            sendErrorResponse(response, ErrorCode.INTERNAL_SERVER_ERROR);
        }
    }

    private void sendErrorResponse(HttpServletResponse response, ErrorCode errorCode) throws IOException {
        response.setStatus(errorCode.getHttpStatus());
        response.setCharacterEncoding("UTF-8");
        response.setContentType("application/json");
        ObjectMapper objectMapper = new ObjectMapper();
        ErrorResponse errorResponse = ErrorResponse.of(errorCode);
        Map<String, ErrorResponse> result = new HashMap<>();
        result.put("result", errorResponse);
        response.getWriter().write(objectMapper.writeValueAsString(result));
    }
}

File: src/main/kotlin/land/leets/global/filter/ExceptionHandleFilter.kt
package land.leets.global.filter

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import land.leets.global.error.ErrorCode
import land.leets.global.error.ErrorResponse
import land.leets.global.error.exception.ServiceException
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException

class ExceptionHandleFilter : OncePerRequestFilter() {

    @Throws(IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        runCatching {
            filterChain.doFilter(request, response)
        }.onFailure { exception ->
            val errorCode = when (exception) {
                is ServiceException -> exception.getErrorCode() // TODO: ServiceException을 Kotlin으로 마이그레이션 후 get 메서드 제거
                else -> ErrorCode.INTERNAL_SERVER_ERROR
            }
            sendErrorResponse(response, errorCode)
        }
    }

    @Throws(IOException::class)
    private fun sendErrorResponse(response: HttpServletResponse, errorCode: ErrorCode) {
        response.apply {
            status = errorCode.getHttpStatus() // TODO: ErrorCode를 Kotlin으로 마이그레이션 후 get 메서드 제거
            characterEncoding = "UTF-8"
            contentType = "application/json"

            val errorResponse = ErrorResponse.of(errorCode)
            val result = mapOf("result" to errorResponse)

            writer.write(ObjectMapper().writeValueAsString(result))
        }
    }
}
